### PR TITLE
Upgrade C++ code to use std::mutex & std::condition_variable

### PR DESCRIPTION
**Pull request by pepone**
This PR removes the remaining usages of IceUtil::Mutex and IceUtil::Monitor, it also includes #1803

### CODE REVIEW

**Review by bernardnormier — APPROVED** [review]

**Review by bernardnormier — APPROVED** [review]
Looks good.

File: cpp/test/Ice/metrics/AllTests.cpp [hunk 1]
```diff
@@ -1144,19 +954,29 @@ allTests(Test::TestHelper* helper, const CommunicatorObserverIPtr& obsv)
...
-            test(false);
-        },
-        [cb](exception_ptr e)
-        {
-            cb->exception(e);
-        });
-    cb->waitForResponse();
+    {
+        std::promise<void> p;
+        metrics->opWithUserExceptionAsync(
```
Comment (externl) [inline cpp/test/Ice/metrics/AllTests.cpp:959]: Why not just use the future API here?

File: cpp/test/Ice/metrics/AllTests.cpp [hunk 2]
```diff
@@ -1175,19 +995,29 @@ allTests(Test::TestHelper* helper, const CommunicatorObserverIPtr& obsv)
...
-            test(false);
-        },
-        [cb](exception_ptr e)
-        {
-            cb->exception(e);
-        });
-    cb->waitForResponse();
+    {
+        std::promise<void> p;
+        metrics->opWithRequestFailedExceptionAsync(
```
Comment (externl) [inline cpp/test/Ice/metrics/AllTests.cpp:1000]: Ditto.

**Review by externl — APPROVED** [review]
Looks good!

File: cpp/test/Ice/metrics/AllTests.cpp (hunk 1 above)
Comment (pepone) [reply cpp/test/Ice/metrics/AllTests.cpp:959]: We test with the 3 APIs

```
  try
    {
        metrics->opWithUserException();
        test(false);
    }
    catch(const Test::UserEx&)
    {
    }

    try
    {
        metrics->opWithUserExceptionAsync().get();
        test(false);
    }
    catch(const Test::UserEx&)
    {
    }
    catch(...)
    {
        test(false);
    }

    {
        std::promise<void> p;
        metrics->opWithUserExceptionAsync(
            []() { test(false); },
            [&p](exception_ptr e)
            {
                try
                {
                    rethrow_exception(e);
                }
                catch(const Test::UserEx&)
                {
                    p.set_value();
                }
                catch(...)
                {
                    test(false);
                }
            });
        p.get_future().wait();
    }
```

Maybe not require, but I didn't change that.  We should probably cleanup this in a future PR!

**Pull request merged**